flink-tests/src/test/java/org/apache/flink/test/runtime/SchedulingLocalRecoveryITCase.java

```diff
@@ -0,0 +1,227 @@
+        configuration.setString(RestOptions.BIND_PORT, "0");
+
+        final long slotIdleTimeout = TIMEOUT;
+        configuration.setLong(JobManagerOptions.SLOT_IDLE_TIMEOUT, slotIdleTimeout);
+
+        configuration.set(TaskManagerOptions.TOTAL_FLINK_MEMORY, MemorySize.parse("64mb"));
+        configuration.set(TaskManagerOptions.FRAMEWORK_HEAP_MEMORY, MemorySize.parse("16mb"));
+        configuration.set(TaskManagerOptions.FRAMEWORK_OFF_HEAP_MEMORY, MemorySize.parse("16mb"));
+
+        final MiniClusterConfiguration miniClusterConfiguration =
+                new MiniClusterConfiguration.Builder()
+                        .setConfiguration(configuration)
+                        .setNumTaskManagers(parallelism)
+                        .setNumSlotsPerTaskManager(1)
+                        .build();
+
+        try (MiniCluster miniCluster = new MiniCluster(miniClusterConfiguration)) {
+            miniCluster.start();
+
+            MiniClusterClient miniClusterClient = new MiniClusterClient(configuration, miniCluster);
+
+            JobGraph jobGraph = createJobGraph(parallelism);
+
+            // wait for the submission to succeed
+            JobID jobId = miniClusterClient.submitJob(jobGraph).get();
+
+            // wait until all tasks running before triggering task failures
+            waitUntilAllVerticesRunning(jobId, miniCluster, TIMEOUT);
+
+            // kill one TM to trigger task failure and remove one existing slot
+            CompletableFuture<Void> terminationFuture = miniCluster.terminateTaskManager(0);
+            terminationFuture.get();
+
+            // restart a taskmanager as a replacement for the killed one
+            miniCluster.startTaskManager();
+
+            // wait until all tasks running again
+            waitUntilAllVerticesRunning(jobId, miniCluster, TIMEOUT);
+
+            ArchivedExecutionGraph graph =
+                    miniCluster.getArchivedExecutionGraph(jobGraph.getJobID()).get();
+
+            miniCluster.cancelJob(jobId).get();
+
+            return graph;
+        }
+    }
+
+    private void waitUntilAllVerticesRunning(
+            JobID jobId, MiniCluster miniCluster, long maxWaitMillis) throws Exception {
+        final Deadline deadline = Deadline.fromNow(Duration.ofMillis(maxWaitMillis));
+
+        boolean checkResult = false;
+        do {
+            AccessExecutionGraph graph = miniCluster.getArchivedExecutionGraph(jobId).get();
+
+            // we need the job status to be RUNNING first to ensure execution vertices are created
+            if (graph.getState() != JobStatus.RUNNING) {
+                continue;
+            }
+
+            checkResult = areAllVerticesRunning(graph);
+
+            if (!checkResult) {
+                Thread.sleep(2L);
+            }
+        } while (!checkResult && deadline.hasTimeLeft());
+
+        if (!checkResult) {
+            throw new TimeoutException("Not all vertices becomes RUNNING in time.");
+        }
+    }
```

Comment:
Maybe we can use `CommonTestUtils.waitForAllTaskRunning`.

Reply:
Thanks for the suggestion! `CommonTestUtils.waitForAllTaskRunning` is good to be used here.
I just looked into `ExecutionGraphTestUtils` and did not notice `CommonTestUtils`.

Reply:
Done.